python-test2.py

```diff
@@ -1,61 +1,83 @@
 import random
 
 class InventoryItem:
+    """Represents an item in inventory with name, quantity, and price."""
+
     def __init__(self, name, quantity, price):
+        if quantity < 0 or price < 0:
+            raise ValueError("Quantity and price must be non-negative.")
         self.name = name
         self.quantity = quantity
         self.price = price
 
     def update_price(self, new_price):
-        if new_price > 0:
-            self.price = new_price
+        """Update the price of the item if the new price is positive."""
+        if new_price < 0:
+            raise ValueError("Price cannot be negative.")
+        self.price = new_price
 
     def update_quantity(self, new_quantity, increment=True):
+        """Update the item's quantity, either incrementing or decrementing."""
         if increment:
             self.quantity += new_quantity
         else:
             self.quantity -= new_quantity
-            if self.quantity < 0:
-                self.quantity = 0
+        if self.quantity < 0:
+            self.quantity = 0
 
 class InventoryManagement:
```

Comment:
Commit Message:
Update llm-review.yml

Commit Diff:
@@ -10,30 +10,32 @@ on:
   issues:
     types: [ opened ]
   issue_comment:
-    types: [ created ]
-  pull_request:
-  pull_request_target:
-    types: [ opened, synchronize, reopened ]
-  pull_request_review_comment:
-    types: [ created ]
   discussion:
-    types: [ created ]
   discussion_comment:
+  pull_request_review_comment:
     types: [ created ]
+  pull_request_target:
+    types: [ opened, synchronize, reopened, edited ]
+  workflow_dispatch:
+    inputs:
+      payload:
+        description: 'The payload of the event that invokes this trigger'
+        required: false
+      event_name:
+        description: 'The name of the event that invokes this trigger'
+        required: false
 
 concurrency:
   group:
     ${{ github.repository }}-${{ github.event.number || github.head_ref ||
     github.sha }}-${{ github.workflow }}-${{ github.event_name }}
-  cancel-in-progress: true
+  cancel-in-progress: ${{ github.event_name == 'pull_request_target' }}
 
 jobs:
   seinesailor:
-    # the github.event.pull_request.head.repo.fork is a boolean, it ensures that the pull_request_target event only
-    # runs from a forked repo
     if: |
-      github.event_name == 'pull_request' ||
-      github.event_name == 'pull_request_target' && github.event.pull_request.head.repo.fork ||
+      github.event_name == 'workflow_dispatch' ||
+      github.event_name == 'pull_request_target' ||
       (contains(github.event.comment.body, '@SeineSailor') && (
         github.event_name == 'issue_comment' || 
         github.event_name == 'pull_request_review_comment' ||
@@ -41,9 +43,27 @@ jobs:
       (github.event_name == 'issues' && contains(github.event.issue.body, '@SeineSailor')) ||
       (github.event_name == 'discussion' && contains(github.event.discussion.body, '@SeineSailor'))
     runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
+      WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
+      OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+      REDIRECT_EVENT_NAME: ${{ github.event.inputs.event_name }}
+      REDIRECT_EVENT_PAYLOAD: ${{ github.event.inputs.payload }}
     steps:
+      - name: Check if redirected
+        id: check_redirect
+        run: |
+          if [ "${{ github.event.inputs.event_name }}" != "" ]; then
+            echo "This is a redirected run"
+            echo "is_redirected=true" >> $GITHUB_ENV
+          else
+            echo "This is the initial run"
+            echo "is_redirected=false" >> $GITHUB_ENV
+          fi
       - name: Check for mention outside of quotes in comments
         id: check_mention
+        if: env.is_redirected == 'false'
         env:
           COMMENT_BODY: ${{ github.event.comment.body }}
         run: |
@@ -69,20 +89,37 @@ jobs:
             fi
           done <<< "$COMMENT_BODY"
           echo "mention_outside_quotes=$mention_outside_quotes" >> $GITHUB_ENV
-
+      - name: Check for empty API KEYs
+        id: check_secret
+        if: env.is_redirected == 'false'
+        run: |
+          if [[ "${{ secrets.IBM_CLOUD_API_KEY }}" == "" 
+          && "${{ secrets.OPENAI_API_KEY }}" == "" 
+          && "${{ github.event_name }}" == "pull_request_review_comment" ]]; then
+            echo "Both IBM_CLOUD_API_KEY and OPENAI_API_KEY are empty, assuming triggered by a forked repo"
+            echo "needs_redirect=true" >> $GITHUB_ENV
+          else
+            echo "needs_redirect=false" >> $GITHUB_ENV
+          fi
+      - name: Trigger Self Workflow in the base repo
+        if: env.needs_redirect == 'true'
+        run: |
+          workflow_filename=$(echo $GITHUB_WORKFLOW_REF | awk -F'@' '{print $1}' | awk -F'/' '{print $NF}')
+          echo "extracted workflow_filename: $workflow_filename"
+          curl -L -X POST -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
+            https://api.github.com/repos/SeineAI/Seine-HelloWorld/actions/workflows/llm-review.yml/dispatches \
+            -d '{"ref": "main", "inputs": {"payload": "${{ toJson(github.event) }}", "event_name": "${{ github.event_name }}"}}'
+        shell: bash
 #      - name: Dump GitHub context for debug
 #        env:
-#          GITHUB_CONTEXT: ${{ toJson(github) }}
+#          EVENT_CONTEXT: ${{ toJson(github) }}
 #        run: |
-#          echo "$GITHUB_CONTEXT"
+#          echo "EVENT_CONTEXT"
       - name: Dump event payload for debug
         run: cat $GITHUB_EVENT_PATH
       - name: Run SeineSailor
+        if: env.needs_redirect != 'true'
         uses: SeineAI/SeineSailor@dev-action-only-do-not-merge-or-delete
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
-          WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
-          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
         with:
           debug: true

@@ -0,0 +1,31 @@
+name: Code Review
+
+permissions:
+  contents: read
+  pull-requests: write
+
+on:
+  pull_request_target:
+    types: [opened, synchronize, reopened]
+  pull_request_review_comment:
+    types: [created]
+
+concurrency:
+  group:
+    ${{ github.repository }}-${{ github.event.number || github.head_ref ||
+    github.sha }}-${{ github.workflow }}-${{ github.event_name ==
+    'pull_request_review_comment' && 'pr_comment' || 'pr' }}
+  cancel-in-progress: ${{ github.event_name != 'pull_request_review_comment' }}
+
+jobs:
+  review:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: coderabbitai/ai-pr-reviewer@latest
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+        with:
+          debug: false
+          review_simple_changes: false
+          review_comment_lgtm: false
\ No newline at end of file

@@ -0,0 +1,89 @@
+class Book:
+    def __init__(self, title, author, ISBN, price, stock):
+        self.title = title
+        self.author = author
+        self.ISBN = ISBN
+        self.price = price
+        self.stock = stock
+
+    def update_stock(self, amount):
+        if amount < 0:
+            print("Error: Stock amount cannot be negative.")
+        else:
+            self.stock += amount
+
+    def sell_book(self, quantity):
+        if quantity <= self.stock:
+            self.stock -= quantity
+            print(f"{quantity} copies of {self.title} sold.")
+        else:
+            print("Not enough stock to sell")
+
+    def set_price(self, new_price):
+        if new_price < 0:
+            print("Error: New price should not be negative.")
+        else:
+            self.price = new_price
+
+class Inventory:
+    def __init__(self):
+        self.books = []
+
+    def add_book(self, book):
+        for b in self.books:
+            if b.ISBN == book.ISBN:
+                print("Book already exists in inventory.")
+                return
+        self.books.append(book)
+
+    def remove_book(self, ISBN):
+        self.books = [book for book in self.books if book.ISBN != ISBN]
+
+    def find_book_by_title(self, title):
+        found_books = [book for book in self.books if title.lower() in book.title.lower()]
+        return found_books
+
+    def find_book_by_author(self, author):
+        found_books = [book for book in self.books if author.lower() in book.author.lower()]
+        return found_books
+
+    def update_stock(self, ISBN, amount):
+        found = False
+        for book in self.books:
+            if book.ISBN == ISBN:
+                book.update_stock(amount)
+                found = True
+        if not found:
+            print("Book not found in inventory.")
+
+    def sell_book(self, ISBN, quantity):
+        for book in self.books:
+            if book.ISBN == ISBN:
+                book.sell_book(quantity)
+                return
+        print("Book not found in inventory.")
+
+    def list_inventory(self):
+        if len(self.books) == 0:
+            print("Inventory is empty.")
+        else:
+            for book in self.books:
+                print(f"Title: {book.title}, Author: {book.author}, ISBN: {book.ISBN}, Price: {book.price}, Stock: {book.stock}")
+
+# Sample usage
+inventory = Inventory()
+book1 = Book("The Great Gatsby", "F. Scott Fitzgerald", "123456789", 20.5, 50)
+book2 = Book("1984", "George Orwell", "987654321", 15.0, 100)
+
+inventory.add_book(book1)
+inventory.add_book(book2)
+
+book3 = Book("To Kill a Mockingbird", "Harper Lee", "123123123", -5, 30)
+inventory.add_book(book3)
+
+inventory.sell_book("123456789", 3)
+inventory.update_stock("987654321", 20)
+
+# Print inventory
+inventory.list_inventory()
+

@@ -0,0 +1,75 @@
+import random
+
+class InventoryItem:
+    def __init__(self, name, quantity, price):
+        self.name = name
+        self.quantity = quantity
+        self.price = price
+
+    def update_price(self, new_price):
+        if new_price > 0:
+            self.price = new_price
+
+    def update_quantity(self, new_quantity, increment=True):
+        if increment:
+            self.quantity += new_quantity
+        else:
+            self.quantity -= new_quantity
+            if self.quantity < 0:
+                self.quantity = 0
+
+class InventoryManagement:
+    def __init__(self):
+        self.items = {}
+
+    def add_item(self, name, quantity, price):
+        if name in self.items:
+            self.items[name].update_quantity(quantity)
+        else:
+            self.items[name] = InventoryItem(name, quantity, price)
+
+    def delete_item(self, name):
+        if name in self.items:
+            del self.items[name]
+
+    def update_item_price(self, name, new_price):
+        if name in self.items:
+            self.items[name].update_price(new_price)
+
+    def update_item_quantity(self, name, quantity, increment=True):
+        if name in self.items:
+            self.items[name].update_quantity(quantity, increment)
+
+    def get_inventory(self):
+        return {name: (item.quantity, item.price) for name, item in self.items.items()}
+
+    def load_items(self, filename):
+        with open(filename, "r") as file:
+            lines = file.readlines()
+            for line in lines:
+                name, quantity, price = line.split(',')
+                self.add_item(name, int(quantity), float(price))
+
+    def save_items(self, filename):
+        with open(filename, "w") as file:
+            for name, item in self.items.items():
+                file.write(f"{name},{item.quantity},{item.price}\n")
+
+def simulate_operations():
+    inventory = InventoryManagement()
+    inventory.add_item("Apple", 100, 0.5)
+    inventory.add_item("Banana", 150, 0.3)
+    inventory.add_item("Carrot", 200, 0.2)
+
+    inventory.update_item_price("Banana", 0.25)
+    inventory.update_item_quantity("Apple", 50, False)
+    inventory.update_item_quantity("Carrot", 50)
+
+    inventory.delete_item("Banana")
+
+    inventory.load_items("inventory.txt")
+    inventory.save_items("inventory_updated.txt")
+
+    print(inventory.get_inventory())
+
+simulate_operations()


Commit Message:
Update llm-review.yml

Commit Diff:
@@ -10,30 +10,32 @@ on:
   issues:
     types: [ opened ]
   issue_comment:
-    types: [ created ]
-  pull_request:
-  pull_request_target:
-    types: [ opened, synchronize, reopened ]
-  pull_request_review_comment:
-    types: [ created ]
   discussion:
-    types: [ created ]
   discussion_comment:
+  pull_request_review_comment:
     types: [ created ]
+  pull_request_target:
+    types: [ opened, synchronize, reopened, edited ]
+  workflow_dispatch:
+    inputs:
+      payload:
+        description: 'The payload of the event that invokes this trigger'
+        required: false
+      event_name:
+        description: 'The name of the event that invokes this trigger'
+        required: false
 
 concurrency:
   group:
     ${{ github.repository }}-${{ github.event.number || github.head_ref ||
     github.sha }}-${{ github.workflow }}-${{ github.event_name }}
-  cancel-in-progress: true
+  cancel-in-progress: ${{ github.event_name == 'pull_request_target' }}
 
 jobs:
   seinesailor:
-    # the github.event.pull_request.head.repo.fork is a boolean, it ensures that the pull_request_target event only
-    # runs from a forked repo
     if: |
-      github.event_name == 'pull_request' ||
-      github.event_name == 'pull_request_target' && github.event.pull_request.head.repo.fork ||
+      github.event_name == 'workflow_dispatch' ||
+      github.event_name == 'pull_request_target' ||
       (contains(github.event.comment.body, '@SeineSailor') && (
         github.event_name == 'issue_comment' || 
         github.event_name == 'pull_request_review_comment' ||
@@ -41,9 +43,27 @@ jobs:
       (github.event_name == 'issues' && contains(github.event.issue.body, '@SeineSailor')) ||
       (github.event_name == 'discussion' && contains(github.event.discussion.body, '@SeineSailor'))
     runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
+      WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
+      OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+      REDIRECT_EVENT_NAME: ${{ github.event.inputs.event_name }}
+      REDIRECT_EVENT_PAYLOAD: ${{ github.event.inputs.payload }}
     steps:
+      - name: Check if redirected
+        id: check_redirect
+        run: |
+          if [ "${{ github.event.inputs.event_name }}" != "" ]; then
+            echo "This is a redirected run"
+            echo "is_redirected=true" >> $GITHUB_ENV
+          else
+            echo "This is the initial run"
+            echo "is_redirected=false" >> $GITHUB_ENV
+          fi
       - name: Check for mention outside of quotes in comments
         id: check_mention
+        if: env.is_redirected == 'false'
         env:
           COMMENT_BODY: ${{ github.event.comment.body }}
         run: |
@@ -69,20 +89,37 @@ jobs:
             fi
           done <<< "$COMMENT_BODY"
           echo "mention_outside_quotes=$mention_outside_quotes" >> $GITHUB_ENV
-
+      - name: Check for empty API KEYs
+        id: check_secret
+        if: env.is_redirected == 'false'
+        run: |
+          if [[ "${{ secrets.IBM_CLOUD_API_KEY }}" == "" 
+          && "${{ secrets.OPENAI_API_KEY }}" == "" 
+          && "${{ github.event_name }}" == "pull_request_review_comment" ]]; then
+            echo "Both IBM_CLOUD_API_KEY and OPENAI_API_KEY are empty, assuming triggered by a forked repo"
+            echo "needs_redirect=true" >> $GITHUB_ENV
+          else
+            echo "needs_redirect=false" >> $GITHUB_ENV
+          fi
+      - name: Trigger Self Workflow in the base repo
+        if: env.needs_redirect == 'true'
+        run: |
+          workflow_filename=$(echo $GITHUB_WORKFLOW_REF | awk -F'@' '{print $1}' | awk -F'/' '{print $NF}')
+          echo "extracted workflow_filename: $workflow_filename"
+          curl -L -X POST -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
+            https://api.github.com/repos/SeineAI/Seine-HelloWorld/actions/workflows/llm-review.yml/dispatches \
+            -d '{"ref": "main", "inputs": {"payload": "${{ toJson(github.event) }}", "event_name": "${{ github.event_name }}"}}'
+        shell: bash
 #      - name: Dump GitHub context for debug
 #        env:
-#          GITHUB_CONTEXT: ${{ toJson(github) }}
+#          EVENT_CONTEXT: ${{ toJson(github) }}
 #        run: |
-#          echo "$GITHUB_CONTEXT"
+#          echo "EVENT_CONTEXT"
       - name: Dump event payload for debug
         run: cat $GITHUB_EVENT_PATH
       - name: Run SeineSailor
+        if: env.needs_redirect != 'true'
         uses: SeineAI/SeineSailor@dev-action-only-do-not-merge-or-delete
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
-          WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
-          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
         with:
           debug: true

@@ -0,0 +1,31 @@
+name: Code Review
+
+permissions:
+  contents: read
+  pull-requests: write
+
+on:
+  pull_request_target:
+    types: [opened, synchronize, reopened]
+  pull_request_review_comment:
+    types: [created]
+
+concurrency:
+  group:
+    ${{ github.repository }}-${{ github.event.number || github.head_ref ||
+    github.sha }}-${{ github.workflow }}-${{ github.event_name ==
+    'pull_request_review_comment' && 'pr_comment' || 'pr' }}
+  cancel-in-progress: ${{ github.event_name != 'pull_request_review_comment' }}
+
+jobs:
+  review:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: coderabbitai/ai-pr-reviewer@latest
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+        with:
+          debug: false
+          review_simple_changes: false
+          review_comment_lgtm: false
\ No newline at end of file

@@ -0,0 +1,89 @@
+class Book:
+    def __init__(self, title, author, ISBN, price, stock):
+        self.title = title
+        self.author = author
+        self.ISBN = ISBN
+        self.price = price
+        self.stock = stock
+
+    def update_stock(self, amount):
+        if amount < 0:
+            print("Error: Stock amount cannot be negative.")
+        else:
+            self.stock += amount
+
+    def sell_book(self, quantity):
+        if quantity <= self.stock:
+            self.stock -= quantity
+            print(f"{quantity} copies of {self.title} sold.")
+        else:
+            print("Not enough stock to sell")
+
+    def set_price(self, new_price):
+        if new_price < 0:
+            print("Error: New price should not be negative.")
+        else:
+            self.price = new_price
+
+class Inventory:
+    def __init__(self):
+        self.books = []
+
+    def add_book(self, book):
+        for b in self.books:
+            if b.ISBN == book.ISBN:
+                print("Book already exists in inventory.")
+                return
+        self.books.append(book)
+
+    def remove_book(self, ISBN):
+        self.books = [book for book in self.books if book.ISBN != ISBN]
+
+    def find_book_by_title(self, title):
+        found_books = [book for book in self.books if title.lower() in book.title.lower()]
+        return found_books
+
+    def find_book_by_author(self, author):
+        found_books = [book for book in self.books if author.lower() in book.author.lower()]
+        return found_books
+
+    def update_stock(self, ISBN, amount):
+        found = False
+        for book in self.books:
+            if book.ISBN == ISBN:
+                book.update_stock(amount)
+                found = True
+        if not found:
+            print("Book not found in inventory.")
+
+    def sell_book(self, ISBN, quantity):
+        for book in self.books:
+            if book.ISBN == ISBN:
+                book.sell_book(quantity)
+                return
+        print("Book not found in inventory.")
+
+    def list_inventory(self):
+        if len(self.books) == 0:
+            print("Inventory is empty.")
+        else:
+            for book in self.books:
+                print(f"Title: {book.title}, Author: {book.author}, ISBN: {book.ISBN}, Price: {book.price}, Stock: {book.stock}")
+
+# Sample usage
+inventory = Inventory()
+book1 = Book("The Great Gatsby", "F. Scott Fitzgerald", "123456789", 20.5, 50)
+book2 = Book("1984", "George Orwell", "987654321", 15.0, 100)
+
+inventory.add_book(book1)
+inventory.add_book(book2)
+
+book3 = Book("To Kill a Mockingbird", "Harper Lee", "123123123", -5, 30)
+inventory.add_book(book3)
+
+inventory.sell_book("123456789", 3)
+inventory.update_stock("987654321", 20)
+
+# Print inventory
+inventory.list_inventory()
+

@@ -0,0 +1,75 @@
+import random
+
+class InventoryItem:
+    def __init__(self, name, quantity, price):
+        self.name = name
+        self.quantity = quantity
+        self.price = price
+
+    def update_price(self, new_price):
+        if new_price > 0:
+            self.price = new_price
+
+    def update_quantity(self, new_quantity, increment=True):
+        if increment:
+            self.quantity += new_quantity
+        else:
+            self.quantity -= new_quantity
+            if self.quantity < 0:
+                self.quantity = 0
+
+class InventoryManagement:
+    def __init__(self):
+        self.items = {}
+
+    def add_item(self, name, quantity, price):
+        if name in self.items:
+            self.items[name].update_quantity(quantity)
+        else:
+            self.items[name] = InventoryItem(name, quantity, price)
+
+    def delete_item(self, name):
+        if name in self.items:
+            del self.items[name]
+
+    def update_item_price(self, name, new_price):
+        if name in self.items:
+            self.items[name].update_price(new_price)
+
+    def update_item_quantity(self, name, quantity, increment=True):
+        if name in self.items:
+            self.items[name].update_quantity(quantity, increment)
+
+    def get_inventory(self):
+        return {name: (item.quantity, item.price) for name, item in self.items.items()}
+
+    def load_items(self, filename):
+        with open(filename, "r") as file:
+            lines = file.readlines()
+            for line in lines:
+                name, quantity, price = line.split(',')
+                self.add_item(name, int(quantity), float(price))
+
+    def save_items(self, filename):
+        with open(filename, "w") as file:
+            for name, item in self.items.items():
+                file.write(f"{name},{item.quantity},{item.price}\n")
+
+def simulate_operations():
+    inventory = InventoryManagement()
+    inventory.add_item("Apple", 100, 0.5)
+    inventory.add_item("Banana", 150, 0.3)
+    inventory.add_item("Carrot", 200, 0.2)
+
+    inventory.update_item_price("Banana", 0.25)
+    inventory.update_item_quantity("Apple", 50, False)
+    inventory.update_item_quantity("Carrot", 50)
+
+    inventory.delete_item("Banana")
+
+    inventory.load_items("inventory.txt")
+    inventory.save_items("inventory_updated.txt")
+
+    print(inventory.get_inventory())
+
+simulate_operations()


Commit Message:
Update llm-review.yml

Commit Diff:
@@ -10,30 +10,32 @@ on:
   issues:
     types: [ opened ]
   issue_comment:
-    types: [ created ]
-  pull_request:
-  pull_request_target:
-    types: [ opened, synchronize, reopened ]
-  pull_request_review_comment:
-    types: [ created ]
   discussion:
-    types: [ created ]
   discussion_comment:
+  pull_request_review_comment:
     types: [ created ]
+  pull_request_target:
+    types: [ opened, synchronize, reopened, edited ]
+  workflow_dispatch:
+    inputs:
+      payload:
+        description: 'The payload of the event that invokes this trigger'
+        required: false
+      event_name:
+        description: 'The name of the event that invokes this trigger'
+        required: false
 
 concurrency:
   group:
     ${{ github.repository }}-${{ github.event.number || github.head_ref ||
     github.sha }}-${{ github.workflow }}-${{ github.event_name }}
-  cancel-in-progress: true
+  cancel-in-progress: ${{ github.event_name == 'pull_request_target' }}
 
 jobs:
   seinesailor:
-    # the github.event.pull_request.head.repo.fork is a boolean, it ensures that the pull_request_target event only
-    # runs from a forked repo
     if: |
-      github.event_name == 'pull_request' ||
-      github.event_name == 'pull_request_target' && github.event.pull_request.head.repo.fork ||
+      github.event_name == 'workflow_dispatch' ||
+      github.event_name == 'pull_request_target' ||
       (contains(github.event.comment.body, '@SeineSailor') && (
         github.event_name == 'issue_comment' || 
         github.event_name == 'pull_request_review_comment' ||
@@ -41,9 +43,27 @@ jobs:
       (github.event_name == 'issues' && contains(github.event.issue.body, '@SeineSailor')) ||
       (github.event_name == 'discussion' && contains(github.event.discussion.body, '@SeineSailor'))
     runs-on: ubuntu-latest
+    env:
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
+      WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
+      OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+      REDIRECT_EVENT_NAME: ${{ github.event.inputs.event_name }}
+      REDIRECT_EVENT_PAYLOAD: ${{ github.event.inputs.payload }}
     steps:
+      - name: Check if redirected
+        id: check_redirect
+        run: |
+          if [ "${{ github.event.inputs.event_name }}" != "" ]; then
+            echo "This is a redirected run"
+            echo "is_redirected=true" >> $GITHUB_ENV
+          else
+            echo "This is the initial run"
+            echo "is_redirected=false" >> $GITHUB_ENV
+          fi
       - name: Check for mention outside of quotes in comments
         id: check_mention
+        if: env.is_redirected == 'false'
         env:
           COMMENT_BODY: ${{ github.event.comment.body }}
         run: |
@@ -69,20 +89,37 @@ jobs:
             fi
           done <<< "$COMMENT_BODY"
           echo "mention_outside_quotes=$mention_outside_quotes" >> $GITHUB_ENV
-
+      - name: Check for empty API KEYs
+        id: check_secret
+        if: env.is_redirected == 'false'
+        run: |
+          if [[ "${{ secrets.IBM_CLOUD_API_KEY }}" == "" 
+          && "${{ secrets.OPENAI_API_KEY }}" == "" 
+          && "${{ github.event_name }}" == "pull_request_review_comment" ]]; then
+            echo "Both IBM_CLOUD_API_KEY and OPENAI_API_KEY are empty, assuming triggered by a forked repo"
+            echo "needs_redirect=true" >> $GITHUB_ENV
+          else
+            echo "needs_redirect=false" >> $GITHUB_ENV
+          fi
+      - name: Trigger Self Workflow in the base repo
+        if: env.needs_redirect == 'true'
+        run: |
+          workflow_filename=$(echo $GITHUB_WORKFLOW_REF | awk -F'@' '{print $1}' | awk -F'/' '{print $NF}')
+          echo "extracted workflow_filename: $workflow_filename"
+          curl -L -X POST -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}" \
+            https://api.github.com/repos/SeineAI/Seine-HelloWorld/actions/workflows/llm-review.yml/dispatches \
+            -d '{"ref": "main", "inputs": {"payload": "${{ toJson(github.event) }}", "event_name": "${{ github.event_name }}"}}'
+        shell: bash
 #      - name: Dump GitHub context for debug
 #        env:
-#          GITHUB_CONTEXT: ${{ toJson(github) }}
+#          EVENT_CONTEXT: ${{ toJson(github) }}
 #        run: |
-#          echo "$GITHUB_CONTEXT"
+#          echo "EVENT_CONTEXT"
       - name: Dump event payload for debug
         run: cat $GITHUB_EVENT_PATH
       - name: Run SeineSailor
+        if: env.needs_redirect != 'true'
         uses: SeineAI/SeineSailor@dev-action-only-do-not-merge-or-delete
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          IBM_CLOUD_API_KEY: ${{ secrets.IBM_CLOUD_API_KEY }}
-          WATSONX_PROJECT_ID: ${{ secrets.WATSONX_PROJECT_ID }}
-          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
         with:
           debug: true

@@ -0,0 +1,31 @@
+name: Code Review
+
+permissions: